server: implement the systemd notification protocol

bdarnell opened the pull request:
Add `--background` flag to `cockroach start` that uses the readiness
protocol to decide when the parent process exits (this is a more
complicated version of the traditional daemonization protocol because Go
doesn't expose the necessary system calls). This can reduce the need to
sleep during test scripts, although until we make `quit` synchronous
we can't eliminate sleeps from reference_test.go.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6268)

<!-- Reviewable:end -->

bdarnell committed:
server: implement the systemd notification protocol

Add `--background` flag to `cockroach start` that uses the readiness
protocol to decide when the parent process exits (this is a more
complicated version of the traditional daemonization protocol because Go
doesn't expose the necessary system calls). This can reduce the need to
sleep during test scripts, although until we make `quit` synchronous
we can't eliminate sleeps from reference_test.go.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[acceptance/reference_test.go, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBO1um0G_GU2E9lGsh:-KGBO1um0G_GU2E9lGsi:-126955565) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/acceptance/reference_test.go#L56)):*
Was there anything wrong with `wait`?

---

*[cli/flags.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBO6-wcFUjkhTAHhqG:-KGBO6-wcFUjkhTAHhqH:1883723750) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/cli/flags.go#L66)):*
s/control/, control/?

---

*[server/server.go, line 378 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBOqv7wbRumHfFdv1Y:-KGBOqv8cas2fnuVeHqJ:-1217215511) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/server/server.go#L378)):*
This would be a reason to exit, wouldn't it?

---

*[util/sdnotify/sdnotify.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBR8zAVSJMIvaG45kh:-KGBR8zAVSJMIvaG45ki:363155563) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L42)):*
Could add that it must only be called once.

---

*[util/sdnotify/sdnotify.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBP2R8-CQTHtsMPHn-:-KGBP2R8-CQTHtsMPHn0:1375892246) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L55)):*
This stuff will break when not run on unix (?). Should it be made to no-op in these cases (for example via a build tag)?

---

*[util/sdnotify/sdnotify.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBP_v8jUmOOoxyoLzm:-KGBP_v9_iMzEeUnLPtQ:296189190) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L85)):*
Does the latest entry win (what happens if I have `NOTIFY_SOCKET` in my env)?

---

*[util/sdnotify/sdnotify.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBRFB1ICCZeXD3P-56:-KGBRFB1ICCZeXD3P-57:981378596) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L106)):*
I didn't see this used.

---

*[util/sdnotify/sdnotify.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBQYa7whEljrXATB5M:-KGBQYa7whEljrXATB5N:153265965) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L113)):*
no removal?

---

*[util/sdnotify/sdnotify.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBQkBXVuRj8cibQZa1:-KGBQkBXVuRj8cibQZa2:848405123) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L141)):*
so technically `s/readyMsg/readyPrefix/`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-:-KGBRRxHl7-ykdBo9Dfr:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCX79qhLSVekLSRcD7:-KGCX79qhLSVekLSRcD8:249158407) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/cli/flags.go#L64)):*
Is that space before `Start` doing anything? I thought leading spaces where trimmed in these messages.

---

*[server/server.go, line 377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCZFkYw3GUd-WE5ERX:-KGCZFkYw3GUd-WE5ERY:-725096430) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/server/server.go#L377)):*
I would think this should be the very last thing to happen in this method. Otherwise, we signal readiness but aren't able to handle grpc-gateway calls.

---

*[util/sdnotify/sdnotify.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCYp7DaFYrgw1Bab8I:-KGCYp7EJ3Y4QXzLEhTI:695170395) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L88)):*
Can combine this an the line above: `if err := cmd.Start(); err != nil {`

---

*[util/sdnotify/sdnotify.go, line 92 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCXp58LPS--6b59uZW:-KGCXp58LPS--6b59uZX:-39350550) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L92)):*
Do you plan to fix this? Seems like no, in which case you can get rid of the `TODO`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-:-KGCZTBgjJTqmVsBT4US:1417088721)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[acceptance/reference_test.go, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBO1um0G_GU2E9lGsh:-KGCgcl1EfY3SBor3EqO:648518897) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/acceptance/reference_test.go#L56)):*
With the move to self-backgrounding rather than `&` on start, there's nothing left for bash to `wait` for.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 4 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[cli/flags.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCX79qhLSVekLSRcD7:-KGDRZuBT9k0YPm9aW94:-1420157938) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/cli/flags.go#L64)):*
Oops, bad `fill-paragraph`. The space is supposed to be a newline (just for consistency with the other usage text; it's technically irrelevant thanks to `wrapText`)

---

*[cli/flags.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBO6-wcFUjkhTAHhqG:-KGDRcjMSQpCQDB1qYED:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/cli/flags.go#L66)):*
Done

---

*[server/server.go, line 377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCZFkYw3GUd-WE5ERX:-KGDSVb82N5ReKlMFwpC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/server/server.go#L377)):*
Done.

---

*[server/server.go, line 378 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBOqv7wbRumHfFdv1Y:-KGDRnnT1nUkYMM_ytx0:1287711919) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/server/server.go#L378)):*
I'm not sure. For comparison, etcd logs and continues in this case.  The worst case of continuing past this error is that systemd will believe that the process never became ready. 

---

*[util/sdnotify/sdnotify.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBR8zAVSJMIvaG45kh:-KGDSl3aW_NEqD-bOqAs:1752489959) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L42)):*
Should, not must. I think it's harmless to call more than once. 

---

*[util/sdnotify/sdnotify.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBP2R8-CQTHtsMPHn-:-KGDSszL4KuPS08gyNSD:2127367176) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L55)):*
It will only run when `$NOTIFY_SOCKET` is set, so it should be reasonably safe on other platforms. We can put it behind a build tag if it turns out to be a problem in practice.

---

*[util/sdnotify/sdnotify.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBP_v8jUmOOoxyoLzm:-KGDVAK81PlA8Anxs-il:-1556004921) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L85)):*
Looks like it's undefined. (`/usr/bin/env` will print both values if there are duplicates; Go will use the first one). I'll change it to filter out duplicates (it's annoying that `os.Setenv` only runs on the singleton environment and there's no way to apply the same processing to a `[]string`).

---

*[util/sdnotify/sdnotify.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCYp7DaFYrgw1Bab8I:-KGDVFuNq3P_Opm0qMdj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L88)):*
Done.

---

*[util/sdnotify/sdnotify.go, line 92 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGCXp58LPS--6b59uZW:-KGDTv-ygrRhI7DY4NVy:33617156) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L92)):*
Removed the TODO (but left the comment). I was thinking this could be published as an independent package in which case this might be worth fixing, but there's no reason to fix it as long as `start --background` is the only use.

---

*[util/sdnotify/sdnotify.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBRFB1ICCZeXD3P-56:-KGDTodlJww_OOeT1xIx:1740281843) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L106)):*
Removed.

---

*[util/sdnotify/sdnotify.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBQYa7whEljrXATB5M:-KGDTmQvBg3L826seBoq:1883512408) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L113)):*
It's removed in `listener.close()`

---

*[util/sdnotify/sdnotify.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268#-KGBQkBXVuRj8cibQZa1:-KGDTxJ66M6WGTWEH6mw:-1114043658) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ab297383cce8d1defeac22e025c0b85c17fcbd/util/sdnotify/sdnotify.go#L141)):*
No, `readyMsg` is the whole message. The `:n` is because `Read()` only writes into a prefix of the supplied slice. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6268)*
<!-- Sent from Reviewable.io -->
